zerostash-files/src/tree.rs

```diff
@@ -0,0 +1,260 @@
+use std::{
+    collections::BTreeMap,
+    sync::{Arc, Mutex},
+};
+
+#[derive(Clone, Serialize, Deserialize, Default, PartialEq, Debug)]
+pub struct File(pub String);
+
+impl File {
+    pub fn new(name: String) -> Self {
+        Self(name)
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub enum Node {
+    File(File),
+    Directory(Arc<Mutex<BTreeMap<String, Node>>>),
+}
+
+impl Default for Node {
+    fn default() -> Self {
+        Self::Directory(Arc::new(Mutex::new(BTreeMap::default())))
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, Default, Debug)]
+pub struct Tree(Arc<Mutex<BTreeMap<String, Node>>>);
```

Comment:
How does this perform when you have something like 100,000 directories with maybe 100,000 files each?

Reply:
Probably not good. Depending on the depth of the tree it gave me stack overflow errors. For reasonable sized trees its probably not too bad